src/software/ai/hl/stp/tactic/attacker/attacker_tactic.cpp

```diff
@@ -38,33 +47,64 @@ bool AttackerTactic::done() const
-        skill_graph_.getNextSkill(tactic_update.robot, *tactic_update.world_ptr);
+    // Update the policy if we are currently executing a skill
+    updatePolicy({world, strategy_});
 
-    if (last_execution_robot == tactic_update.robot.id())
+    current_skill_.reset();
+}
+
+void AttackerTactic::updatePrimitive(const TacticUpdate& tactic_update, bool reset_fsm)
+{
+    // If the robot executing the current skill is done, or if there is no current skill
+    // assigned, we need to assign a new skill
+    if (!current_skill_ || (last_execution_robot == tactic_update.robot.id() &&
+                            current_skill_->done(tactic_update.robot.id())))
     {
-        if (current_skill_ == nullptr || current_skill_->done(tactic_update.robot.id()))
-        {
-            current_skill_ = next_skill;
-            current_skill_->reset(tactic_update.robot);
-            skill_graph_.extendSequence(current_skill_);
-        }
-
-        current_skill_->updatePrimitive(tactic_update.robot, tactic_update.world_ptr,
-                                        tactic_update.set_primitive);
+        AttackerMdpState attacker_mdp_state{tactic_update.world_ptr, strategy_};
+
+        // Update the policy if we completed executing a skill
+        updatePolicy(attacker_mdp_state);
+
+        // Update ActionSelectionStrategy hyperparameters
+        action_selection_strategy_->setEpsilon(
+            strategy_->getAiConfig().attacker_tactic_config().action_selection_epsilon());
+
+        // Select the next skill to execute according to the policy
+        auto action    = policy_.selectAction(attacker_mdp_state);
+        current_skill_ = createSkillFromAttackerMdpAction(action, strategy_);
+
+        gameplay_monitor_.startStepObservation(tactic_update.world_ptr);
     }
-    else
+
+    if (reset_fsm)
     {
-        next_skill->reset(tactic_update.robot);
-        next_skill->updatePrimitive(tactic_update.robot, tactic_update.world_ptr,
-                                    tactic_update.set_primitive);
+        current_skill_->reset(tactic_update.robot);
     }
+
+    current_skill_->updatePrimitive(tactic_update.robot, tactic_update.world_ptr,
+                                    tactic_update.set_primitive);
 }
 
-void AttackerTactic::evaluate(double score)
+void AttackerTactic::updatePolicy(const AttackerMdpState& attacker_mdp_state)
 {
-    skill_graph_.scoreSequence(score);
-    current_skill_ = nullptr;
+    // Only update the policy if we previously selected a skill
+    if (current_skill_)
+    {
+        const TbotsProto::AttackerTacticConfig& attacker_config =
+            strategy_->getAiConfig().attacker_tactic_config();
+
+        double reward =
+            gameplay_monitor_.endStepObservation(attacker_mdp_state.world_ptr);
+
+        // Update Q-function hyperparameters
+        q_function_->setLearningRate(attacker_config.learning_rate());
+        q_function_->setDiscountFactor(attacker_config.discount_factor());
```

Comment:
it looks like `q_function_` is initialized with these values, what's your reasoning for updating them during runtime?

Reply:
if we change learning_rate or discount_factor dynamic param during runtime, the q-function won't be automatically updated with the new param values